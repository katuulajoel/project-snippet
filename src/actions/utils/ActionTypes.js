export const STATUS_APPROVED = "approved";
export const STATUS_PENDING = "pending";

export const DOC_TYPE_OTHER = "other";

// Dashboard action types
export const FETCH_NOTIFICATIONS_START = "FETCH_NOTIFICATIONS_START";
export const FETCH_NOTIFICATIONS_SUCCESS = "FETCH_NOTIFICATIONS_SUCCESS";
export const FETCH_NOTIFICATIONS_FAILED = "FETCH_NOTIFICATIONS_FAILED";
export const DELETE_NOTIFICATIONS_START = "DELETE_NOTIFICATIONS_START";
export const DELETE_NOTIFICATIONS_SUCCESS = "DELETE_NOTIFICATIONS_SUCCESS";
export const DELETE_NOTIFICATIONS_FAILED = "DELETE_NOTIFICATIONS_FAILED";
export const CREATE_NOTIFICATION_LOG_START = "CREATE_NOTIFICATION_LOG_START";
export const CREATE_NOTIFICATION_LOG_SUCCESS =
  "CREATE_NOTIFICATION_LOG_SUCCESS";
export const CREATE_NOTIFICATION_LOG_FAILED = "CREATE_NOTIFICATION_LOG_FAILED";

// Invoices action types
export const CREATE_INVOICE_START = "CREATE_INVOICE_START";
export const CREATE_INVOICE_SUCCESS = "CREATE_INVOICE_SUCCESS";
export const CREATE_INVOICE_FAILED = "CREATE_INVOICE_FAILED";
export const CREATE_INVOICE_BATCH_START = "CREATE_INVOICE_BATCH_START";
export const CREATE_INVOICE_BATCH_SUCCESS = "CREATE_INVOICE_BATCH_SUCCESS";
export const CREATE_INVOICE_BATCH_FAILED = "CREATE_INVOICE_BATCH_FAILED";
export const LIST_INVOICES_START = "LIST_INVOICES_START";
export const LIST_INVOICES_SUCCESS = "LIST_INVOICES_SUCCESS";
export const LIST_INVOICES_FAILED = "LIST_INVOICES_FAILED";
export const SEARCH_INVOICES_START = "SEARCH_INVOICES_START";
export const SEARCH_INVOICES_SUCCESS = "SEARCH_INVOICES_SUCCESS";
export const SEARCH_INVOICES_FAILED = "SEARCH_INVOICES_FAILED";
export const RETRIEVE_INVOICE_START = "RETRIEVE_INVOICE_START";
export const RETRIEVE_INVOICE_SUCCESS = "RETRIEVE_INVOICE_SUCCESS";
export const RETRIEVE_INVOICE_FAILED = "RETRIEVE_INVOICE_FAILED";
export const UPDATE_INVOICE_START = "UPDATE_INVOICE_START";
export const UPDATE_INVOICE_SUCCESS = "UPDATE_INVOICE_SUCCESS";
export const UPDATE_INVOICE_FAILED = "UPDATE_INVOICE_FAILED";
export const LIST_MORE_INVOICES_START = "LIST_MORE_INVOICES_START";
export const LIST_MORE_INVOICES_SUCCESS = "LIST_MORE_INVOICES_SUCCESS";
export const LIST_MORE_INVOICES_FAILED = "LIST_MORE_INVOICES_FAILED";
export const SEARCH_MORE_INVOICES_START = "SEARCH_MORE_INVOICES_START";
export const SEARCH_MORE_INVOICES_SUCCESS = "SEARCH_MORE_INVOICES_SUCCESS";
export const SEARCH_MORE_INVOICES_FAILED = "SEARCH_MORE_INVOICES_FAILED";
export const DELETE_INVOICE_START = "DELETE_INVOICE_START";
export const DELETE_INVOICE_SUCCESS = "DELETE_INVOICE_SUCCESS";
export const DELETE_INVOICE_FAILED = "DELETE_INVOICE_FAILED";
export const PAY_INVOICE_START = "PAY_INVOICE_START";
export const PAY_INVOICE_SUCCESS = "PAY_INVOICE_SUCCESS";
export const PAY_INVOICE_FAILED = "PAY_INVOICE_FAILED";
export const ARCHIVE_INVOICE_SUCCESS = "ARCHIVE_INVOICE_SUCCESS";
export const ARCHIVE_INVOICE_START = "ARCHIVE_INVOICE_START";
export const ARCHIVE_INVOICE_FAILED = "ARCHIVE_INVOICE_FAILED";
export const GENERATE_INVOICE_SUCCESS = "GENERATE_INVOICE_SUCCESS";
export const GENERATE_INVOICE_START = "GENERATE_INVOICE_START";
export const GENERATE_INVOICE_FAILED = "GENERATE_INVOICE_FAILED";
export const BULK_ACTION_START = "BULK_ACTION_START";
export const BULK_ACTION_FAILED = "BULK_ACTION_FAILED";
export const BULK_ACTION_SUCCESS = "BULK_ACTION_SUCCESS";
export const DOWNLOAD_INVOICE_CSV_STARTED = "DOWNLOAD_INVOICE_CSV_STARTED";
export const DOWNLOAD_INVOICE_CSV_FAILED = "DOWNLOAD_INVOICE_CSV_FAILED";
export const DOWNLOAD_INVOICE_CSV_SUCCESS = "DOWNLOAD_INVOICE_CSV_SUCCESS";
export const INVOICE_SUMMARY_START = "INVOICE_SUMMARY_START";
export const INVOICE_SUMMARY_SUCCESS = "INVOICE_SUMMARY_SUCCESS";
export const INVOICE_SUMMARY_FAILED = "INVOICE_SUMMARY_FAILED";
export const SEARCH_PROJECTS_START = "SEARCH_PROJECTS_START";
export const SEARCH_PROJECTS_SUCCESS = "SEARCH_PROJECTS_SUCCESS";
export const SEARCH_PROJECTS_FAILED = "SEARCH_PROJECTS_FAILED";
export const SEARCH_MORE_PROJECTS_START = "SEARCH_MORE_PROJECTS_START";
export const SEARCH_MORE_PROJECTS_SUCCESS = "SEARCH_MORE_PROJECTS_SUCCESS";
export const SEARCH_MORE_PROJECTS_FAILED = "SEARCH_MORE_PROJECTS_FAILED";

// Projects Actions types
export const FETCH_PROJECT_START = "FETCH_PROJECT_START";
export const FETCH_PROJECT_SUCCESS = "FETCH_PROJECT_SUCCESS";
export const FETCH_PROJECT_FAILED = "FETCH_PROJECT_FAILED";
// Invites action types
export const GET_PENDING_INVITES_START = "GET_PENDING_INVITES_START";
export const GET_PENDING_INVITES_SUCCESS = "GET_PENDING_INVITES_SUCCESS";
export const GET_PENDING_INVITES_FAILED = "GET_PENDING_INVITES_FAILED";
export const GET_MORE_PENDING_INVITES_START = "GET_MORE_PENDING_INVITES_START";
export const GET_MORE_PENDING_INVITES_SUCCESS =
  "GET_MORE_PENDING_INVITES_SUCCESS";
export const GET_MORE_PENDING_INVITES_FAILED =
  "GET_MORE_PENDING_INVITES_FAILED";
export const INVITE_START = "INVITE_START";
export const INVITE_SUCCESS = "INVITE_SUCCESS";
export const INVITE_FAILED = "INVITE_FAILED";
export const DELETE_INVITE_START = "DELETE_INVITE_START";
export const DELETE_INVITE_SUCCESS = "DELETE_INVITE_SUCCESS";
export const DELETE_INVITE_FAILED = "DELETE_INVITE_FAILED";
