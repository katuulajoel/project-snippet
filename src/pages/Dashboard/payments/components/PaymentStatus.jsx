/* -------------------------------------------------------------------------- */
/*                            External dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React from "react";

export const BatchStatus = (props) => {
  const { batch } = props;
  if (
    (new Date(`${batch.due_at}`) < new Date() &&
      batch.status === "approved" &&
      !batch.paid) ||
    (!batch.paid && batch.status === "approved")
  ) {
    return <span className="pending">Processing</span>;
  }
  if (new Date(`${batch.due_at}`) < new Date() && !batch.paid) {
    return <span className="overdue">Overdue</span>;
  }
  if (batch.paid) {
    return <span className="completed">Paid</span>;
  }
  return <span className="pending">Pending</span>;
};

BatchStatus.propTypes = {
  batch: PropTypes.shape({
    due_at: PropTypes.string,
    paid: PropTypes.bool,
    status: PropTypes.string,
  }),
};

const PaymentStatus = (props) => {
  const { invoice } = props;

  if (new Date(`${invoice.due_at}`) < new Date() && !invoice.paid) {
    return <span className="overdue">Overdue</span>;
  }
  if (invoice.paid) {
    return <span className="completed">Paid</span>;
  }
  return <span className="pending">Pending</span>;
};

PaymentStatus.propTypes = {
  invoice: PropTypes.shape({
    due_at: PropTypes.string,
    paid: PropTypes.bool,
  }),
};

export default PaymentStatus;
