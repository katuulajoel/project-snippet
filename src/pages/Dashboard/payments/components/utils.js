import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";

export const GENERATE_INVOICE_ACTION = "GENERATE_INVOICE_ACTION";
export const PAY_ACTION = "PAY_ACTION";
export const EDIT_ACTION = "EDIT_ACTION";
export const DELETE_ACTION = "DELETE_ACTION";
export const MARK_AS_PAID_ACTION = "MARK_AS_PAID_ACTION";
export const ARCHIVE_ACTION = "ARCHIVE_ACTION";

export const showAction = (action, invoice) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case EDIT_ACTION:
    case DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    default:
      return false;
  }
};

// eslint-disable-next-line no-unused-vars
export const performAction = (action, invoice, props) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION: {
      // onGenerateInvoice(invoice.id, props);
      break;
    }
    case PAY_ACTION: {
      /* openModal(<MakePaymentModal invoice={invoice} />, "Make Payment", true, {
        className: "modal-report modal-make-payment",
      }); */
      break;
    }
    case EDIT_ACTION: {
      // onUpdateInvoice(invoice, props);
      break;
    }
    case DELETE_ACTION: {
      // onDeleteInvoice(invoice.id, props);
      break;
    }
    case MARK_AS_PAID_ACTION: {
      // onMarkPaid(invoice.id, props);
      break;
    }
    case ARCHIVE_ACTION: {
      // onMarkArchived(invoice.id, props);
      break;
    }
  }
};
