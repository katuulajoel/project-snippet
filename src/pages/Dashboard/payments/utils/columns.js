// import { isPayAdmin } from "../../../../components/utils/auth";

//TODO: check if use is pay admin to active batch actions, add ...isPayAdmin() && on line 6

export const getTableColumns = (filter) => [
  ...(!(filter === "archived" || filter === "paid")
    ? [
        {
          Header: " ",
          accessor: "batch_action",
        },
      ]
    : []),
  {
    Header: "Date Created",
    accessor: "created_at",
  },
  {
    Header: "Client / Project / Payment Title",
    accessor: "title",
  },
  {
    Header: "Invoice No.",
    accessor: "invoice",
  },
  {
    Header: "Amount",
    accessor: "amount",
  },
  ...(filter !== "paid"
    ? [
        {
          Header: "Due Date",
          accessor: "due_date",
        },
        {
          Header: "Status",
          accessor: "status",
        },
        ...(filter !== "archived"
          ? [
              {
                Header: "",
                accessor: "actions",
              },
            ]
          : []),
      ]
    : []),
];
