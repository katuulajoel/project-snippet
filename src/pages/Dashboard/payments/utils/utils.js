import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";
import * as actions from "./constant"; //TODO: Load just actions needed
import {
  approvePayout,
  generateInvoice,
  markAsArchived,
  markAsPaid,
} from "./paymentActions";

function sumInvoiceAmounts(invoices) {
  return invoices
    .map((invoice) => {
      return invoice.amount || 0;
    })
    .reduce((total, number) => {
      return total + Math.round(number * 100) / 100;
    }, 0);
}

/**
 * Group invoices with the same reference into one batch object
 * @param {*} invoices list of invoices
 * @returns returns array of batch objects
 */
export function batchInvoices(invoices) {
  let batchedInvoices = {};
  invoices?.forEach((invoice) => {
    batchedInvoices[invoice.batch_ref] = [
      ...(batchedInvoices[invoice.batch_ref] || []),
      invoice,
    ];
  });

  let batchObjects = [];
  Object.keys(batchedInvoices).map((ref) => {
    let batchObj = {
      ...batchedInvoices[ref][0],
      invoices: batchedInvoices[ref],
    };
    batchObj.ref = ref;
    batchObj.amount = sumInvoiceAmounts(batchedInvoices[ref]);
    batchObjects.push(batchObj);
  });
  return batchObjects;
}

/**
 * Determine whether to show action or not depending on logged in User
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} invoice invoice if payments action or invoices if payouts action
 * @returns {boolean} false | true
 */
export const showAction = (action, invoice) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case actions.PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case actions.EDIT_ACTION:
    case actions.DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case actions.MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case actions.APPROVE_BATCH_ACTION:
      return isPayAdmin() && invoice.status !== "approved";
    case actions.ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    case actions.EDIT_ACTION_BATCH:
    case actions.DELETE_ACTION_BATCH:
      return (
        isPayAdmin() &&
        !invoice.project.archived &&
        !invoice.status !== "approved" &&
        !invoice.paid
      );
  }
};

/**
 * Determine what function to call depending on the action that been triggered
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} data invoice if payments action or invoices if payouts action
 */
export const performAction = (action, data) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION: {
      generateInvoice(data.id);
      break;
    }
    case actions.MARK_AS_PAID_ACTION: {
      markAsPaid(data.id);
      break;
    }
    case actions.ARCHIVE_ACTION: {
      markAsArchived(data.id);
      break;
    }
    case actions.APPROVE_BATCH_ACTION: {
      approvePayout(data);
      break;
    }
  }
};

/**
 * filters to be used as query params
 * @param {*} filter : selected payment status;
 * @returns object keys value pairs
 */
export const getPaymentsFilters = (filter) => {
  switch (filter) {
    case "paid":
      return { paid: "True" };
    case "pending":
      return { paid: "False", overdue: "False" };
    case "overdue":
      return { overdue: "True", paid: "False" };
    case "archived":
      return { archived: "True" };
    default:
      return {};
  }
};
