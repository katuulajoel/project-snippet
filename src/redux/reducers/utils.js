export const getIds = (items) => {
  return Array.from(
    new Set(
      items.map((item) => {
        return item.id;
      })
    )
  );
};

export const reduceUser = (state, user, profile, company) => {
  state = state || {};
  user = user || {};
  profile = { ...state.profile, ...user.profile, ...profile };
  company = { ...state.company, ...user.company, ...company };
  [
    ["name", "company"],
    ["bio", "company_bio"],
    ["website", "website"],
    ["vat_number", "vat_number"],
    ["reg_no", "company_reg_no"],
    ["ref_no", "reference_number"],
    ["skills", "skills"],
    ["country", "country"],
    ["city", "city"],
    ["street", "street"],
    ["plot_number", "plot_number"],
    ["postal_code", "postal_code"],
    ["postal_address", "postal_address"],
    ["tel_number", "phone_number"],
  ].forEach((item) => {
    let companyField = item[0],
      profileField = item[1];
    company[companyField] = company[companyField] || profile[profileField];
  });
  return { ...state, ...user, profile, company };
};
