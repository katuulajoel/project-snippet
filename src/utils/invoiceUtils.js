import PropTypes from "prop-types";
import React from "react";
import moment from "moment";
import {
  archiveInvoice,
  generateInvoice as generateInvoiceAction,
  updateInvoice,
  downloadInoicesCsv,
  getInvoiceSummary,
  bulkAction,
} from "../redux/actions/InvoiceActions";
import DateRangeForm from "../components/DateRangeForm";
import ModalHeader from "../components/ModalHeader";
import { openConfirm, openModal } from "./modals";
import store from "../redux/store";
import { isPayAdmin, isClient, isPayAdminOrPM } from "./auth";
import * as actions from "../configs/constants/invoiceConstants";

export const Header = (props) => (
  <ModalHeader
    {...props}
    style={{ paddingBottom: "8px" }}
    options={{ title: props.title }}
  />
);

Header.propTypes = {
  title: PropTypes.string,
};

//TODO: check if use is pay admin to active batch actions, add ...isPayAdmin() && on line 6
export const getTableColumns = (filter, project = null) => [
  ...(!(filter === "archived" || filter === "paid")
    ? [
        {
          Header: " ",
          accessor: "batch_action",
        },
      ]
    : []),
  {
    Header: "Date Created",
    accessor: "created_at",
  },
  {
    Header: `${!project ? "Client / Project / Payment " : ""}Title`,
    accessor: "title",
  },
  {
    Header: "Invoice No.",
    accessor: "invoice",
  },
  {
    Header: "Amount",
    accessor: "amount",
  },
  ...(filter !== "paid"
    ? [
        {
          Header: "Due Date",
          accessor: "due_date",
        },
        {
          Header: "Status",
          accessor: "status",
        },
        ...(filter !== "archived"
          ? [
              {
                Header: "",
                accessor: "actions",
              },
            ]
          : []),
      ]
    : []),
];

/**
 * Create data source to be use in react table
 * @param {array} data Array of invoice objects
 * @returns array of objects
 */
export const tableData = (data) => [
  ...data?.map((invoice) => {
    let row = {
      batch_action: invoice,
      created_at: invoice.created_at,
      title: { title: invoice.title, project: invoice.project },
      invoice: {
        id: invoice.id,
        number: invoice.number,
        paid: invoice.paid,
        finalized: invoice.finalized,
        last_sent_at: invoice.last_sent_at,
      },
      amount: {
        total_amount: invoice.total_amount,
        amount: invoice.amount,
        type: invoice.type,
      },
      due_date: { due_at: invoice.due_at, paid: invoice.paid },
      status: {
        id: invoice.id,
        due_at: invoice.due_at,
        paid: invoice.paid,
      },
      actions: { ...invoice },
    };
    return row;
  }),
];

function sumInvoiceAmounts(invoices) {
  return invoices
    .map((invoice) => {
      return invoice.amount || 0;
    })
    .reduce((total, number) => {
      return total + Math.round(number * 100) / 100;
    }, 0);
}

/**
 * Group invoices with the same reference into one batch object
 * @param {*} invoices list of invoices
 * @returns returns array of batch objects
 */
export function batchInvoices(invoices) {
  let batchedInvoices = {};
  invoices?.forEach((invoice) => {
    batchedInvoices[invoice.batch_ref] = [
      ...(batchedInvoices[invoice.batch_ref] || []),
      invoice,
    ];
  });

  let batchObjects = [];
  Object.keys(batchedInvoices).map((ref) => {
    let batchObj = {
      ...batchedInvoices[ref][0],
      invoices: batchedInvoices[ref],
    };
    batchObj.ref = ref;
    batchObj.amount = sumInvoiceAmounts(batchedInvoices[ref]);
    batchObjects.push(batchObj);
  });
  return batchObjects;
}

/**
 * Determine whether to show action or not depending on logged in User
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} invoice invoice if payments action or invoices if payouts action
 * @returns {boolean} false | true
 */
export const showAction = (action, invoice) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case actions.PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case actions.EDIT_ACTION:
    case actions.DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case actions.MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case actions.APPROVE_BATCH_ACTION:
      return isPayAdmin() && invoice.status !== "approved";
    case actions.ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    case actions.EDIT_ACTION_BATCH:
    case actions.DELETE_ACTION_BATCH:
      return (
        isPayAdmin() &&
        !invoice.project.archived &&
        !invoice.status !== "approved" &&
        !invoice.paid
      );
  }
};

/**
 * Determine what function to call depending on the action that been triggered
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} data invoice if payments action or invoices if payouts action
 */
export const performAction = (action, data) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION: {
      generateInvoice(data.id);
      break;
    }
    case actions.MARK_AS_PAID_ACTION: {
      markAsPaid(data.id);
      break;
    }
    case actions.ARCHIVE_ACTION: {
      markAsArchived(data.id);
      break;
    }
    case actions.APPROVE_BATCH_ACTION: {
      approvePayout(data);
      break;
    }
  }
};

/**
 * filters to be used as query params
 * @param {*} filter : selected payment status;
 * @returns object keys value pairs
 */
export const getPaymentsFilters = (filter) => {
  switch (filter) {
    case "paid":
      return { paid: "True" };
    case "pending":
      return { paid: "False", overdue: "False" };
    case "overdue":
      return { overdue: "True", paid: "False" };
    case "archived":
      return { archived: "True" };
    default:
      return {};
  }
};

export function generateInvoice(invoiceId) {
  openConfirm({
    message: "Are you sure you want to generate an invoice for this payment?",
    // title: <Header />,
  }).then(() => {
    store.dispatch(generateInvoiceAction(invoiceId));
  });
}

export function markAsPaid(invoiceId) {
  openConfirm({
    message: "Are you sure you want to mark this invoice as paid?",
    title: "Mark as Paid",
  }).then(() => {
    store.dispatch(updateInvoice(invoiceId, { paid: true }));
  });
}

export function markAsArchived(invoiceId) {
  openConfirm({
    message: "Are you sure you want to archive this invoice?",
    title: "Archive invoice",
  }).then(() => {
    store.dispatch(archiveInvoice(invoiceId));
  });
}

export function approvePayout(invoices) {
  openConfirm({
    message: "Are you sure you want to approve this payout?",
    options: { ok: "Approve", cancel: "Cancel" },
    title: "Approve payout",
  }).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { status: "approved" }));
    });
  });
}

export function bulkMarkAsPaid(invoices) {
  openConfirm({
    message: "Are you sure you want to mark all selected as paid?",
    title: "Bulk mark as Paid",
  }).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { paid: true }));
    });
  });
}

export function bulkGenerateInvoice(invoices) {
  openConfirm({
    message: "Are you sure you want to generate invoices for all selected?",
    title: "Bulk generate invoices",
  }).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(generateInvoiceAction(invoice.id));
    });
  });
}

export function bulkDeleteInvoice(invoices) {
  openConfirm({
    message: "Are you sure you want to delete all selected?",
    title: "Bulk delete invoices",
  }).then(() => {
    store.dispatch(bulkAction(invoices, "delete"));
  });
}

export function bulkArchiveInvoice(invoices) {
  openConfirm({
    message: "Are you sure you want to archive all selected?",
    title: "Bulk archive invoices",
  }).then(() => {
    store.dispatch(bulkAction(invoices, "archive"));
  });
}

export function downloadCsv(filter, type, project = null) {
  openModal({
    body: <DateRangeForm id={`export-form`} />,
    options: {
      className: "modal-payments",
      ok: `Download CSV`,
      form: {
        type: "submit",
        form: `export-form`,
      },
    },
    title: `Export ${type}`,
  }).then((data) => {
    if (data) {
      let params = {};
      if (project) {
        params["project"] = project;
      }

      switch (filter) {
        case "archived":
          params["archived"] = "True";
          params["paid"] = "False";
          break;
        case "all":
          params["archived"] = "False";
          break;
        case "pending":
          params["archived"] = "False";
          params["paid"] = "False";
          params["overdue"] = "False";
          break;
        case "overdue":
          params["archived"] = "False";
          params["paid"] = "False";
          params["overdue"] = "True";
          break;
        case "paid":
          params["archived"] = "False";
          params["paid"] = "True";
      }

      store.dispatch(
        downloadInoicesCsv({
          start: data.start,
          end: data.end,
          type: type === "Payments" ? "sale" : "purchase",
          ...params,
        })
      );
    }
  });
}

export function filterPaymentSummaries(
  { start, end },
  type,
  setSummariesRange
) {
  openModal({
    body: (
      <DateRangeForm
        id={`fiter-payment`}
        defaultStart={start}
        defaultEnd={end}
        maxdate={new Date()}
        message={"Select date range for filter"}
      />
    ),
    options: {
      className: "modal-payments",
      ok: `Submit`,
      form: {
        type: "submit",
        form: `fiter-payment`,
      },
    },
    title: `Filter Total ${type === "Payments" ? "Payments" : "Payouts"}`,
  }).then((data) => {
    if (data) {
      const START = `${moment(data.start).format(
        moment.HTML5_FMT.DATE
      )}T00:00:00`;
      const END = `${moment(data.end).format(moment.HTML5_FMT.DATE)}T23:59:59`;

      setSummariesRange({
        start: START,
        end: END,
      });

      store.dispatch(
        getInvoiceSummary({
          min_date: START,
          max_date: END,
          type: type === "in" ? "sale" : "purchase",
        })
      );
    }
  });
}
