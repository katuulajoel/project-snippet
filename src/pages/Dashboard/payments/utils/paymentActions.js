import React from "react";
import {
  archiveInvoice,
  generateInvoice as generateInvoiceAction,
  updateInvoice,
} from "../../../../actions/InvoiceActions";
import ModalHeader from "../../../../components/ModalHeader";
import { openConfirm } from "../../../../components/utils/modals";
import store from "../../../../store";

export function generateInvoice(invoiceId) {
  openConfirm(
    <div>Are you sure you want to generate an invoice for this payment?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Generate Invoice" }}
    />
  ).then(() => {
    store.dispatch(generateInvoiceAction(invoiceId));
  });
}

export function markAsPaid(invoiceId) {
  openConfirm(
    <div>Are you sure you want to mark this invoice as paid?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Mark as Paid" }}
    />
  ).then(() => {
    store.dispatch(updateInvoice(invoiceId, { paid: true }));
  });
}

export function markAsArchived(invoiceId) {
  openConfirm(
    <div>Are you sure you want to archive this invoice?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Archive invoice" }}
    />
  ).then(() => {
    store.dispatch(archiveInvoice(invoiceId));
  });
}

export function approvePayout(invoices) {
  openConfirm(
    <div>Are you sure you want to approve this payout?</div>,
    null,
    true,
    { ok: "Approve", cancel: "Cancel" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Approve payout" }}
    />
  ).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { status: "approved" }));
    });
  });
}
