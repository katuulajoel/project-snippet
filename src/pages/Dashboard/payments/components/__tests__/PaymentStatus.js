/* eslint-disable no-unused-vars */
import React from "react";
import PaymentStatus, { BatchStatus } from "../PaymentStatus";
import renderer from "react-test-renderer";
import * as utils from "../../../../../utils/invoiceUtils";
import mount from "enzyme/build/mount";

const invoice = {
  due_at: "2021-06-09T23:59:59.999999",
  is_overdue: true,
  status: "pending",
  paid: false,
  finalized: true,
  archived: false,
  created_at: "2021-05-26T18:05:21.072944",
};

describe("Payment Status tests", () => {
  it("Should match snapshot test for payments", () => {
    jest.spyOn(utils, "showAction").mockReturnValueOnce(true);
    const tree = renderer.create(<PaymentStatus invoice={invoice} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should match snapshot test for payouts batch", () => {
    jest.spyOn(utils, "showAction").mockReturnValueOnce(true);
    const tree = renderer.create(<BatchStatus batch={invoice} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render the right payment status", () => {
    let current = new Date();
    let previous = new Date(current.getTime() - 86400000); // + 1 day in ms
    const wrapper = mount(
      <PaymentStatus
        invoice={{ due_at: previous.toISOString(), paid: false }}
      />
    );

    let status = wrapper.find("span");
    expect(wrapper.text().includes("Overdue")).toBe(true);
    wrapper.setProps({
      invoice: { due_at: previous.toISOString(), paid: true },
    });

    let next = new Date(current.getTime() + 86400000); // + 1 day in ms
    wrapper.setProps({
      invoice: { due_at: next.toISOString(), paid: false },
    });
    expect(wrapper.text().includes("Pending")).toBe(true);
  });

  it("should render the right batch status", () => {
    let current = new Date();
    let previous = new Date(current.getTime() - 86400000); // + 1 day in ms
    const wrapper = mount(
      <BatchStatus batch={{ due_at: previous.toISOString(), paid: false }} />
    );

    let status = wrapper.find("span");
    expect(wrapper.text().includes("Overdue")).toBe(true);
    wrapper.setProps({
      batch: { due_at: previous.toISOString(), paid: true },
    });

    wrapper.setProps({
      batch: {
        due_at: previous.toISOString(),
        paid: false,
        status: "approved",
      },
    });
    expect(wrapper.text().includes("Processing")).toBe(true);

    let next = new Date(current.getTime() + 86400000); // + 1 day in ms
    wrapper.setProps({
      batch: { due_at: next.toISOString(), paid: false },
    });
    expect(wrapper.text().includes("Pending")).toBe(true);
  });
});
