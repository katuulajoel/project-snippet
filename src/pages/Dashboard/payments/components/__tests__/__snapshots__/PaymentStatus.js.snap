// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Status tests Should match snapshot test for payments 1`] = `
<span
  className="overdue"
>
  Overdue
</span>
`;

exports[`Payment Status tests Should match snapshot test for payouts batch 1`] = `
<span
  className="overdue"
>
  Overdue
</span>
`;
